# 05.Data Consistency checks

## 01.Importing Libraries

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

## 02.Importing Data

In [2]:
path = r'E:\Aarthi Sarathi\Career Foundry\Exercise 4\Instacart Basket Analsis'
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Original Data', 'orders_wrangled.csv'), index_col = False)

## 03.Mixed Type Data

In [3]:
#creating DataFrame
df_test = pd.DataFrame()
df_test['mix'] = ['a', 'b', 1, True]
df_test.head()

mix
0     a
1     b
2     1
3  True

In [4]:
# Check for any mixed-type columns
for col in df_test.columns.tolist():
    weird = (df_test[col].map(type) != df_test[col].iloc[0].__class__).any()
    if weird:
        print(col)

mix


In [5]:
# Convert column's data from numeric to string
df_test['mix'] = df_test['mix'].astype('str')
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   mix     4 non-null      object
dtypes: object(1)
memory usage: 164.0+ bytes


In [6]:
df_test['mix'].dtype

dtype('O')

## 04. Missing Values

In [7]:
# Find missing values (takes sum of Trues in a  column)
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [8]:
# Create subset of the dataframe that contains the nulls
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [9]:
# Create subset of the dataframe that contains the notnulls
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]
df_prods_clean.shape

(49677, 5)

## 05.Finding Duplicates

In [10]:
##Finding Duplicates
df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [11]:
##Addressing Duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()
df_prods_clean_no_dups.shape

(49672, 5)

Observation - 5 coulmns are duplicated , after dropping duplicated columns there only (49672, 5)

## 06.Exporting

In [12]:
df_prods_clean_no_dups.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'products_checked.csv'))

# Exercise 4.5 Task

## Step 2 Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [13]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       order_day_of_week  order_hour_of_day  days_since_prior_order  
count       3.421083e+06       3.421083e+06            3.214874e+06  
mean        2.776219e+00       1.345202e+01            1.111484e+01  
std         2.046829e+00       4.226088e+00            9.206737e+00  
min         0.000000e+00       0.000000e+00            0.000000e+00  
25%         1.000000e+00       1.000000e+01            4.000000e+00  
50%         3.000000e+00       1.300000e+01            7.000000e+00  
75%         5.000000e+00       1.600000e+01            1.500000e+01  
max         6.000000e+00       2.300000e+01            3.000000e+01

Lets look at the values for 'order_number', 'orders_day_of_week', and 'days_since_prior_order': 
1.'order_number' has a maximum value of 100, which seems very high depending on the time frame
2.'orders_day_of_week' has a minimum of '0' and a maximum of '6' suggesting the days are zero-indexed (0 for Monday or Sunday)
3.There are many missing values in the 'days_since_prior_order' column, as its count is lower than the other columns´. 

## Step 3 and 4 Check for mixed-type data in your df_ords dataframe.
If you find mixed-type data, fix it The column in question should contain observations of a single data type.

In [14]:
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
else :
      print("No mixed columns")

No mixed columns


No mixed type coulmns found

## Step 5 and 6: Run a check for missing values in your df_ords dataframe.
## Address the missing values using an appropriate method.

In [15]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
order_day_of_week              0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [16]:
# creating df only for Null values
df_nan_ords = df_ords[(df_ords['days_since_prior_order']).isnull() == True]
df_nan_ords

Unnamed: 0  order_id  user_id  order_number  order_day_of_week  \
0                 0   2539329        1             1                  2   
11               11   2168274        2             1                  2   
26               26   1374495        3             1                  1   
39               39   3343014        4             1                  6   
45               45   2717275        5             1                  3   
...             ...       ...      ...           ...                ...   
3420930     3420930    969311   206205             1                  4   
3420934     3420934   3189322   206206             1                  3   
3421002     3421002   2166133   206207             1                  6   
3421019     3421019   2227043   206208             1                  1   
3421069     3421069   3154581   206209             1                  3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 7 columns]

In [17]:
#checking frequency of order_number in df_nan_ords
df_nan_ords['order_number'].value_counts()

order_number
1    206209
Name: count, dtype: int64

In [18]:
#checking frequency of order_number in df_ords
df_ords['order_number'].value_counts()

order_number
1      206209
2      206209
3      206209
4      206209
5      182223
        ...  
96       1592
97       1525
98       1471
99       1421
100      1374
Name: count, Length: 100, dtype: int64

#### Removing missing coulmns would result in loss of information so first finding The frequency of missing values (206,209) is the same as the frequency of order_number=1.And Creating a new column that flags new orders (True or False).

In [19]:
# Create a function to determine if the order in question is a first order or not
def is_first_order(row):
    if row['order_number'] == 1:
        return True
    else:
        return False

In [20]:
# Apply the is_first_order function to every row of the df_ords dataframe
df_ords.apply(lambda row: is_first_order(row), axis=1)

0           True
1          False
2          False
3          False
4          False
           ...  
3421078    False
3421079    False
3421080    False
3421081    False
3421082    False
Length: 3421083, dtype: bool

In [21]:
# Create a new column in df_ords to flag the reason for missing values
df_ords['first_order']=df_ords.apply(lambda row: is_first_order(row), axis=1)

In [22]:
# Final check for the frequency of first_order=True
df_ords['first_order'].value_counts()

first_order
False    3214874
True      206209
Name: count, dtype: int64

## Step 7 and 8 Run a check for duplicate values in your df_ords data.
## Address the duplicates using an appropriate method

In [23]:
#ceated data frame to store duplicate orders
df_ords_dup = df_ords[df_ords.duplicated()]
df_ords_dup

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, order_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

## There is no duplicate orders found

### Exporting orders 

In [24]:


df_ords.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_cleaned.csv'))